hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/clean/CleanPlanner.java

```diff
@@ -432,4 +458,47 @@ private boolean isFileSliceNeededForPendingCompaction(FileSlice fileSlice) {
   private boolean isFileGroupInPendingCompaction(HoodieFileGroup fg) {
     return fgIdToPendingCompactionOperations.containsKey(fg.getFileGroupId());
   }
+
+  public boolean isDropPartition(Option<HoodieInstant> instantToRetain) {
+    try {
+      if (instantToRetain.isPresent() && HoodieTimeline.REPLACE_COMMIT_ACTION.equals(instantToRetain.get().getAction())) {
+        HoodieReplaceCommitMetadata replaceCommitMetadata = HoodieReplaceCommitMetadata.fromBytes(
+            hoodieTable.getActiveTimeline().getInstantDetails(instantToRetain.get()).get(), HoodieReplaceCommitMetadata.class);
+
+        if (replaceCommitMetadata != null
+            && WriteOperationType.DELETE_PARTITION.equals(replaceCommitMetadata.getOperationType())) {
+          return true;
+        }
+      }
+    } catch (Exception e) {
+      throw new HoodieException("Failed to get commit metadata", e);
+    }
+    return false;
+  }
+
+  public List<String> getDropPartitions(Option<HoodieInstant> instantToRetain) {
+    try {
+      if (!instantToRetain.isPresent()) {
+        LOG.info("No earliest commit to retain. No need to scan partitions !!");
+        return Collections.emptyList();
+      }
+
+      if (HoodieTimeline.REPLACE_COMMIT_ACTION.equals(instantToRetain.get().getAction())) {
```

Comment:
I am not sure if we can map earliest instant to retain to clean up partitions to delete. 

for eg:
C1, C2(delete partition), C3, C4. 

lets say after C4, earliest commit to retain is C2: this essentially means that cleaner has to clean all data files for all commits < C2. and not touch anything pertaining to C2. but here wrt getDropPartition, we are breaking that and triggering deletion of partitions from C2. 

or am I missing something?

Reply:
well, I see what you mean, I'll check again.